src/nupic/math/topology.py

```diff
@@ -0,0 +1,154 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+import itertools
+import numpy
+
+
+def coordinatesFromIndex(index, dimensions):
+  """
+  Translate an index into coordinates, using the given coordinate system.
+
+  Similar to numpy.unravel_index.
+
+  @param index (int)
+  The index of the point. The coordinates are expressed as a single index by
+  using the dimensions as a mixed radix definition. For example, in dimensions
+  42x10, the point [1, 4] is index 1*420 + 4*10 = 460.
+
+  @param dimensions (list of ints)
+  The coordinate system.
+
+  @returns
+  A list of coordinates of length len(dimensions).
+  """
+  coordinates = [0] * len(dimensions)
+
+  shifted = index
+  for i in xrange(len(dimensions) - 1, 0, -1):
+    coordinates[i] = shifted % dimensions[i]
+    shifted = shifted / dimensions[i]
+
+  coordinates[0] = shifted
+
+  return coordinates
+
+
+def indexFromCoordinates(coordinates, dimensions):
+  """
+  Translate coordinates into an index, using the given coordinate system.
+
+  Similar to numpy.ravel_multi_index.
+
+  @param coordinates (list of ints)
+  A list of coordinates of length dimensions.size().
+
+  @param dimensions (list of ints
+  The coordinate system.
+
+  @returns
+  The index of the point. The coordinates are expressed as a single index by
+  using the dimensions as a mixed radix definition. For example, in dimensions
+  42x10, the point [1, 4] is index 1*420 + 4*10 = 460.
+  """
+  index = 0
+  for i, dimension in enumerate(dimensions):
+    index *= dimension
+    index += coordinates[i]
+
+  return index
+
+
+def neighborhood(centerIndex, radius, dimensions):
```

Comment:
I went ahead and added this extensibility to the SP by moving all calls to `neighborhood` to `_getInputNeighborhood` and `_getColumnNeighborhood` on the SP. This is much easier here in Python because there's no static typing, and of course because in Python we're returning simple lists, not a magical iterator class.